src/Microsoft.DotNet.Cli.Utils/Telemetry.cs

```diff
@@ -0,0 +1,117 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.ApplicationInsights;
+using Microsoft.Extensions.PlatformAbstractions;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    public class Telemetry : ITelemetry
+    {
+        private static bool _isInitialized = false;
+        private static TelemetryClient _client = null;
+
+        private static Dictionary<string, string> _commonProperties = null;
+        private static Dictionary<string, double> _commonMeasurements = null;
+
+        private const string InstrumentationKey = "74cc1c9e-3e6e-4d05-b3fc-dde9101d0254";
+        private const string TelemetryOptout = "DOTNET_CLI_TELEMETRY_OPTOUT";
+        private const string OSVersion = "OS Version";
+        private const string OSPlatform = "OS Platform";
+        private const string RuntimeId = "Runtime Id";
+        private const string ProductVersion = "Product Version";
+
+        public Telemetry()
+        {
+            bool optout = Env.GetEnvironmentVariableAsBool(TelemetryOptout);
+
+            if (optout)
+            {
+                return;
+            }
+
+            try
+            {
+                _client = new TelemetryClient();
+                _client.InstrumentationKey = InstrumentationKey;
+                _client.Context.Session.Id = Guid.NewGuid().ToString();
+
+                var runtimeEnvironment = PlatformServices.Default.Runtime;
+                _client.Context.Device.OperatingSystem = runtimeEnvironment.OperatingSystem;
+
+                _commonProperties = new Dictionary<string, string>();
+                _commonProperties.Add(OSVersion, runtimeEnvironment.OperatingSystemVersion);
+                _commonProperties.Add(OSPlatform, runtimeEnvironment.OperatingSystemPlatform.ToString());
+                _commonProperties.Add(RuntimeId, runtimeEnvironment.GetRuntimeIdentifier());
+                _commonProperties.Add(ProductVersion, Product.Version);
+                _commonMeasurements = new Dictionary<string, double>();
+
+                _isInitialized = true;
+            }
+            catch (Exception)
+            {
+                // we dont want to fail the tool if telemetry fais. We should be able to detect abnormalities from data 
+                // at the server end
```

Comment:
I made a comment in the previous pull request that I didn't see a response on. "I agree with you that this should not fail and have a negative effect on what the end user is trying to achieve. That being said, do you think it would make sense to Debug.Fail or something like that so that there is a better chance issues will be caught during development in case something is wrong?"